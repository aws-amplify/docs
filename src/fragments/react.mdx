## What we'll build

In this tutorial, we'll use React to build a photo sharing app that lets users upload photos to a shared photo gallery. We'll use a GraphQL API (AWS AppSync) to access data stored in a NoSQL database (Amazon DynamoDB). In addition, we'll demonstrate how to authenticate users, communicate with our API, manage photo uploads, and set up continuous deployment and hosting.

The app includes the following views:

- A login view
- A photo gallery view
- A photo upload view

## Prerequisites

- [Install Node](https://nodejs.org/en/)
- This tutorial assumes that you're familiar with both JavaScript/ES6 and React. If you need to brush up on React, we recommend going through the [official tutorial](https://reactjs.org/tutorial/tutorial.html).