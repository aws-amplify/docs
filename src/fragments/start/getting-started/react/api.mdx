Now that you’ve created and configured a React app and initialized a new Amplify project, you can add a feature. The first feature you will add is an API.

The Amplify CLI supports creating and interacting with two types of API categories: REST and GraphQL.

The API you will be creating in this step is a GraphQL API using AWS AppSync (a managed GraphQL service) and the database will be Amazon DynamoDB (a NoSQL database).

## Create a GraphQL API and database

Add a [GraphQL API](https://docs.aws.amazon.com/appsync/latest/devguide/designing-a-graphql-api.html) to your app and automatically provision a database by running the following command from the root of your application directory:

```bash
amplify add api
```

Accept the **default values** which are highlighted below:

```console
? Please select from one of the below mentioned services:
# GraphQL
? Provide API name:
# myapi
? Choose the default authorization type for the API:
# API Key
? Enter a description for the API key:
# demo
? After how many days from now the API key should expire:
# 7 (or your preferred expiration)
? Do you want to configure advanced settings for the GraphQL API:
# No
? Do you have an annotated GraphQL schema?
# No
? Choose a schema template:
# Single object with fields (e.g., “Todo” with ID, name, description)
? Do you want to edit the schema now?
# Yes
```

The CLI should open this GraphQL schema in your text editor.

__amplify/backend/api/myapi/schema.graphql__

```graphql
type Todo @model {
  id: ID!
  name: String!
  description: String
}
```

The schema generated is for a Todo app. You'll notice a directive on the `Todo` type of `@model`. This directive is part of the [GraphQL transform](/cli/graphql-transformer/model) library of Amplify. 

The GraphQL Transform Library provides custom directives you can use in your schema that allow you to do things like define data models, set up authentication and authorization rules, configure serverless functions as resolvers, and more.

A type decorated with the `@model` directive will scaffold out the database table for the type (Todo table), the schema for CRUD (create, read, update, delete) and list operations, and the GraphQL resolvers needed to make everything work together.

From the command line, press __enter__ to accept the schema and continue to the next steps.

### Deploying the API

To deploy this backend, run the `push` command:

```bash
amplify push
```

```console
? Are you sure you want to continue? Y

# You will be walked through the following questions for GraphQL code generation
? Do you want to generate code for your newly created GraphQL API? Y
? Choose the code generation language target: javascript
? Enter the file name pattern of graphql queries, mutations and subscriptions: src/graphql/**/*.js
? Do you want to generate/update all possible GraphQL operations - queries, mutations and subscriptions? Y
? Enter maximum statement depth [increase from default if your schema is deeply nested]: 2
```

Now the API is live and you can start interacting with it!

The API you have deployed is for a Todo app, including operations for creating, reading, updating, deleting, and listing todos.

Next, run the following command to check Amplify's status:

```bash
amplify status
```

This will give us the current status of the Amplify project, including the current environment, any categories that have been created, and what state those categories are in. It should look similar to this:

```console
Current Environment: dev

| Category | Resource name | Operation | Provider plugin   |
| -------- | ------------- | --------- | ----------------- |
| Api      | myapi         | No Change | awscloudformation |
```

To view the GraphQL API in the AppSync console at any time, run the following command:

```bash
amplify console api
```

To view your entire app in the Amplify console at any time, run the following command:

```bash
amplify console
```

### (Optional) Test your API

To test this out locally, you can run the `mock` command.

> If you'd like to go ahead and connect the front end, you can [jump to the next step](#connect-frontend-to-api).

```bash
amplify mock api
```

*Note:* `amplify mock api` requires Java.

```console
# If you have not already deployed you API, you will be walked through the following steps for GraphQL code generation
? Choose the code generation language target: javascript (or preferred target)
? Enter the file name pattern of graphql queries, mutations and subscriptions: src/graphql/**/*.js
? Do you want to generate/update all possible GraphQL operations - queries, mutations and subscriptions: Yes
? Enter maximum statement depth [increase from default if your schema is deeply nested] 2
```

This will open the GraphiQL explorer on a local port. From the test environment you can try out different operations locally, like queries and mutations.

Try running a couple of mutations locally and then querying for the todos:

```graphql
mutation createTodo {
  createTodo(input: {
    name: "Build an API"
    description: "Build a serverless API with Amplify and GraphQL"
  }) {
    id
    name
    description
  }
}

query listTodos {
  listTodos {
    items {
      id
      description
      name
    }
  }
}
```

## Connect frontend to API

In this section you will create a way to list and create todos from the React application. To do this, you will create a form with a button to create todos as well as a way to fetch and render the list of todos.

Open __src/App.js__ and replace it with the following code:

```javascript
/* src/App.js */
import React, { useEffect, useState } from 'react'
import Amplify, { API, graphqlOperation } from 'aws-amplify'
import { createTodo } from './graphql/mutations'
import { listTodos } from './graphql/queries'

import awsExports from "./aws-exports";
Amplify.configure(awsExports);

const initialState = { name: '', description: '' }

const App = () => {
  const [formState, setFormState] = useState(initialState)
  const [todos, setTodos] = useState([])

  useEffect(() => {
    fetchTodos()
  }, [])

  function setInput(key, value) {
    setFormState({ ...formState, [key]: value })
  }

  async function fetchTodos() {
    try {
      const todoData = await API.graphql(graphqlOperation(listTodos))
      const todos = todoData.data.listTodos.items
      setTodos(todos)
    } catch (err) { console.log('error fetching todos') }
  }

  async function addTodo() {
    try {
      if (!formState.name || !formState.description) return
      const todo = { ...formState }
      setTodos([...todos, todo])
      setFormState(initialState)
      await API.graphql(graphqlOperation(createTodo, {input: todo}))
    } catch (err) {
      console.log('error creating todo:', err)
    }
  }

  return (
    <div style={styles.container}>
      <h2>Amplify Todos</h2>
      <input
        onChange={event => setInput('name', event.target.value)}
        style={styles.input}
        value={formState.name}
        placeholder="Name"
      />
      <input
        onChange={event => setInput('description', event.target.value)}
        style={styles.input}
        value={formState.description}
        placeholder="Description"
      />
      <button style={styles.button} onClick={addTodo}>Create Todo</button>
      {
        todos.map((todo, index) => (
          <div key={todo.id ? todo.id : index} style={styles.todo}>
            <p style={styles.todoName}>{todo.name}</p>
            <p style={styles.todoDescription}>{todo.description}</p>
          </div>
        ))
      }
    </div>
  )
}

const styles = {
  container: { width: 400, margin: '0 auto', display: 'flex', flexDirection: 'column', justifyContent: 'center', padding: 20 },
  todo: {  marginBottom: 15 },
  input: { border: 'none', backgroundColor: '#ddd', marginBottom: 10, padding: 8, fontSize: 18 },
  todoName: { fontSize: 20, fontWeight: 'bold' },
  todoDescription: { marginBottom: 0 },
  button: { backgroundColor: 'black', color: 'white', outline: 'none', fontSize: 18, padding: '12px 0px' }
}

export default App
```

Let's walk through some of the functions:

__useEffect__ - When the component loads, the `useEffect` hook is called and it invokes the `fetchTodos` function.

__fetchTodos__ - Uses the Amplify `API` category to call the AppSync GraphQL API with the `listTodos` query. Once the data is returned, the items array is passed in to the `setTodos` function to update the local state.

__addTodo__ - Uses the Amplify `API` category to call the AppSync GraphQL API with the `createTodo` mutation. A difference between the `listTodos` query and the `createTodo` mutation is that `createTodo` accepts an argument containing the variables needed for the mutation.

## Run locally

Next, run the app and you should see the form rendered to the screen and be able to create and view the list of todos:

```bash
npm start
```
