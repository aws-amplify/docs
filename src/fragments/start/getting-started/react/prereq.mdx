- This tutorial assumes that you're familiar with both JavaScript/ES6 and React. If you need to brush up on React, we recommend going through the [official tutorial](https://reactjs.org/tutorial/tutorial.html).