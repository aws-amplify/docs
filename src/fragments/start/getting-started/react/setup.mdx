## Create a new React App

To get started, first create a new React app, and then install and use the Amplify CLI to start adding backend capabilities to your app.

From your projects directory, run the following command and respond the prompts:

```bash
npm create vite@latest
✔ Project name: react-amplified
✔ Select a framework: › React
✔ Select a variant: › TypeScript
```

This creates a new React TypeScript app in a directory called **react-amplified**. You can then run the following commands to switch into the new **react-amplified** directory, install the project dependencies, and run the app by using the following commands:

```bash
cd react-amplified
npm install
npm run dev
```

This runs a development server and allows you to see the output generated by the build. You can see the running app by navigating to [http://localhost:5173](http://localhost:5173).

## Initialize a new backend

With the app running, it's time to set up Amplify and create the backend services to support the app.

Open a new terminal. From the **react-amplified** directory, run:

```bash
amplify init
```

When you initialize Amplify you'll be prompted for some information about the app, with the option to accept recommended values:

```console
? Enter a name for the project reactamplified
The following configuration will be applied:

?Project information
| Name:  reactamplified
| Environment: dev
| Default editor: Visual Studio Code
| App type: javascript
| Javascript framework: react
| Source Directory Path: src
| Distribution Directory Path: build
| Build Command: npm run-script build
| Start Command: npm run-script start

? Initialize the project with the above configuration? Yes
Using default provider  awscloudformation
? Select the authentication method you want to use: AWS profile

...

? Please choose the profile you want to use default
```

> Amplify CLI will infer the proper configuration based on the type of project Amplify is being initialized in where possible. Here it detected Create React App and provided the proper configuration for type of app, framework, source, distribution, build, and start options.

When you initialize a new Amplify project, a few things happen:

- It creates a top level directory called **amplify** that stores your backend definition. During the tutorial you'll add capabilities such as a GraphQL API and authentication. As you add features, the **amplify** folder will grow with infrastructure-as-code templates that define your backend stack. Infrastructure-as-code is a best practice way to create a replicable backend stack.
- It creates a file called **amplifyconfiguration.json** in the **src** directory that holds all the configuration for the services you create with Amplify. This is how the Amplify client is able to get the necessary information about your backend services.
- It modifies the **.gitignore** file, adding some generated files to the ignore list
- A cloud project is created for you in the AWS Amplify Console that can be accessed by running `amplify console`. The Console provides a list of backend environments, deep links to provisioned resources per Amplify category, status of recent deployments, and instructions on how to promote, clone, pull, and delete backend resources

## Install Amplify Libraries

The `aws-amplify` package is the main library for working with Amplify Libraries in your projects:

```bash
npm install aws-amplify
```

## Set up frontend

Next, configure Amplify so it can interact with backend services.

Open **src/index.js** and add the following code below the last import:

```javascript
import { Amplify } from 'aws-amplify';
import amplifyconfig from './amplifyconfiguration.json';
Amplify.configure(amplifyconfig);
```

And that's all it takes to configure Amplify. As you add or remove categories and make updates to your backend configuration using the Amplify CLI, the configuration in **aws-exports.js** will update automatically.

Now that your React app is set up and Amplify is initialized, you can add an API in the next step.
