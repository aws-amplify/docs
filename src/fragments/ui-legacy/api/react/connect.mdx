The API category provides React components for working with GraphQL data using the Amplify GraphQL client. 

The `<Connect/>` component is used to execute a GraphQL query or mutation. You can execute GraphQL queries by passing your queries in `query` or `mutation` attributes:

## Queries

```javascript
import React, { Component } from 'react';
import Amplify, { graphqlOperation }  from "aws-amplify";
import { Connect } from "aws-amplify-react";

import * as queries from './graphql/queries';
import * as subscriptions from './graphql/subscriptions';

class App extends Component {

    render() {

        const ListView = ({ todos }) => (
            <div>
                <h3>All Todos</h3>
                <ul>
                    {todos.map(todo => <li key={todo.id}>{todo.name} ({todo.id})</li>)}
                </ul>
            </div>
        );

        return (
            <Connect query={graphqlOperation(queries.listTodos)}>
                {({ data: { listTodos }, loading, errors }) => {
                    if (errors) return (<h3>Error</h3>);
                    if (loading || !listTodos) return (<h3>Loading...</h3>);
                    return (<ListView todos={listTodos.items} /> );
                }}
            </Connect>
        )
    }
} 

export default App;

```

## Subscription

Also, you can use the `subscription` and `onSubscriptionMsg` attributes to enable subscriptions:

```javascript

<Connect
    query={graphqlOperation(queries.listTodos)}
    subscription={graphqlOperation(subscriptions.onCreateTodo)}
    onSubscriptionMsg={(prev, { onCreateTodo }) => {
        console.log(onCreateTodo);
        return prev; 
    }}
>
    {({ data: { listTodos }, loading, error }) => {
        if (error) return (<h3>Error</h3>);
        if (loading || !listTodos) return (<h3>Loading...</h3>);
        return (<ListView todos={listTodos ? listTodos.items : []} />);
    }}
 </Connect>

```
## Mutations

For mutations, a `mutation` function needs to be provided with the `Connect` component. A `mutation` returns a promise that resolves with the result of the GraphQL mutation.

```jsx
import React, { Component } from 'react';
import * as mutations from './graphql/mutations';
import * as queries from './graphql/queries';
import * as subscriptions from './graphql/subscriptions';

class AddTodo extends Component {
  constructor(props) {
    super(props);
    this.submit = this.submit.bind(this);
    this.state = {
        name: '',
        description: '',
    };
  }

  handleChange(name, event) {
      this.setState({ [name]: event.target.value });
  }

  async submit() {
    const { onCreate } = this.props;
    const input = {
      name: this.state.name,
      description: this.state.description
    }
    console.log(input);

    try {
    	await onCreate({input})
    } catch (err) {
    	console.error(err);
    }

  }

  render(){
    return (
        <div>
            <input
                name="name"
                placeholder="name"
                onChange={(event) => { this.handleChange('name', event)}}
            />
            <input
                name="description"
                placeholder="description"
                onChange={(event) => { this.handleChange('description', event)}}
            />
            <button onClick={this.submit}>
                Add
            </button>
        </div>
    );
  }
}

class App extends Component {
  render() {

    const ListView = ({ todos }) => (
      <div>
          <h3>All Todos</h3>
          <ul>
            {todos.map(todo => <li key={todo.id}>{todo.name}</li>)}
          </ul>
      </div>
    )

    return (
      <div className="App">
        <Connect mutation={graphqlOperation(mutations.createTodo)}>
          {({mutation}) => (
            <AddTodo onCreate={mutation} />
          )}
        </Connect>

        <Connect query={graphqlOperation(queries.listTodos)}
          subscription={graphqlOperation(subscriptions.onCreateTodo)}
          onSubscriptionMsg={(prev, {onCreateTodo}) => {
              console.log('Subscription data:', onCreateTodo)
              return prev;
            }
          }>
        {({ data: { listTodos }, loading, error }) => {
          if (error) return <h3>Error</h3>;
          if (loading || !listTodos) return <h3>Loading...</h3>;
            return (<ListView todos={listTodos.items} />);
        }}
        </Connect>
      </div>

    );
  }
}
```
