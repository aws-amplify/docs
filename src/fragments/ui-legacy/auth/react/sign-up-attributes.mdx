<table>
  <thead>
    <tr>
      <th>Attribute</th>
      <th>Type</th>
      <th>Description</th>
      <th>Default</th>
      <th>Required</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td data-column="Attribute">header</td>
      <td data-column="Type">string</td>
      <td data-column="Description">the component header</td>
      <td data-column="Default">'Create a new account'</td>
      <td data-column="Required">no</td>
    </tr>
    <tr>
      <td data-column="Attribute">signUpFields</td>
      <td data-column="Type">array</td>
      <td data-column="Description">see below</td>
      <td data-column="Default">see below</td>
      <td data-column="Required">no</td>
    </tr>
    <tr>
      <td data-column="Attribute">defaultCountryCode</td>
      <td data-column="Type">string</td>
      <td data-column="Description">
        the preselected value in the country code dropdown
      </td>
      <td data-column="Default">'1'</td>
      <td data-column="Required">no</td>
    </tr>
    <tr>
      <td data-column="Attribute">hideAllDefaults</td>
      <td data-column="Type">boolean</td>
      <td data-column="Description">
        determines whether all default signup fields are to be hidden. This
        works in conjunction with the signUpFields attribute
      </td>
      <td data-column="Default">
        if there is no signUpFields attribute, defaults to false
      </td>
      <td data-column="Required">no</td>
    </tr>
    <tr>
      <td data-column="Attribute">hiddenDefaults</td>
      <td data-column="Type">array</td>
      <td data-column="Description">
        determines whether particular default fields are hidden
      </td>
      <td data-column="Default">
        N/A (possible values include 'username', 'password', 'phone_number', and
        'email')
      </td>
      <td data-column="Required">no</td>
    </tr>
  </tbody>
</table>