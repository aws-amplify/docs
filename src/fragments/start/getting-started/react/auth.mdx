The next feature you will be adding is authentication.

## Authentication with Amplify

Amplify uses [Amazon Cognito](https://aws.amazon.com/cognito/) as the main authentication provider. Amazon Cognito is a robust user directory service that handles user registration, authentication, account recovery & other operations. In this tutorial, you'll learn how to add authentication to your application using Amazon Cognito and username/password login.

## Create authentication service

To add authentication to your app, run this command:

```bash
amplify add auth
```

Select the defaults for the following prompts:

```console
? Do you want to use the default authentication and security configuration? Default configuration
Warning: you will not be able to edit these selections.
? How do you want users to be able to sign in? Username
? Do you want to configure advanced settings? No, I am done.
```

To deploy the service, run the `push` command:

```bash
amplify push
```

Now, the authentication service has been deployed and you can start using it. To view the deployed services in your project at any time, go to Amplify Console by running the following command:

```bash
amplify console
```

## Create login UI

Creating a login flow can be quite difficult and time consuming to get right. Luckily, Amplify UI has an [authentication](https://ui.docs.amplify.aws/react/components/authenticator) component that provides an entire authentication flow for you, using the configuration you specified in **amplifyconfiguration.json**.

### Install Amplify UI

The `@aws-amplify/ui-react` package includes React specific UI components you'll use to build your app. Install it with this:

```bash
npm install @aws-amplify/ui-react
```

### Add the Amplify UI Authenticator component

Open **src/App.jsx** and make the following changes:

1. Import the `withAuthenticator` component:

```javascript
import { withAuthenticator, Button, Heading } from '@aws-amplify/ui-react';
import '@aws-amplify/ui-react/styles.css';
```

2. Pass `{signOut, user}` to the `App` component. (For Vite's default React configuration, you'll also need to add "PropTypes" to validate the `signOut` and `user` props):

```javascript
/* src/App.jsx */
import PropTypes from 'prop-types'
// ... other imports

const App = ({ signOut, user }) => {
  // ...
}

App.propTypes = {
  signOut: PropTypes.func,
  user: PropTypes.object
}
```

3. Add this heading and button block to the top of your `App` component:

```javascript
return (
  <div style={styles.container}>
    <Heading level={1}>Hello {user.username}</Heading>
    <Button onClick={signOut}>Sign out</Button>
    <h2>Amplify Todos</h2>
    ...
  </div>
);
```

4. Lastly, wrap your `App` export with the `withAuthenticator` Amplify UI component:

```javascript
export default withAuthenticator(App);
```

Run the app to see the new authentication flow protecting the app:

```bash
npm run dev
```

Now you should see the app load with an authentication flow allowing users to sign up and sign in. Here's all the code below:

```javascript
/* src/App.js */
import { useEffect, useState } from 'react';
import PropTypes from 'prop-types'
import { Amplify } from 'aws-amplify';
import { generateClient } from 'aws-amplify/api';
import { createTodo } from './graphql/mutations';
import { listTodos } from './graphql/queries';
import { withAuthenticator, Button, Heading } from '@aws-amplify/ui-react';
import '@aws-amplify/ui-react/styles.css';

import awsExports from './amplifyconfiguration.json';
Amplify.configure(awsExports);

const initialState = { name: '', description: '' };
const client = generateClient();

const App = ({ signOut, user }) => {
  const [formState, setFormState] = useState(initialState);
  const [todos, setTodos] = useState([]);

  useEffect(() => {
    fetchTodos();
  }, []);

  function setInput(key, value) {
    setFormState({ ...formState, [key]: value });
  }

  async function fetchTodos() {
    try {
      const todoData = await client.graphql({
        query: listTodos
      });
      const todos = todoData.data.listTodos.items;
      setTodos(todos);
    } catch (err) {
      console.log('error fetching todos');
    }
  }

  async function addTodo() {
    try {
      if (!formState.name || !formState.description) return;
      const todo = { ...formState };
      setTodos([...todos, todo]);
      setFormState(initialState);
      await client.graphql({
        query: createTodo,
        variables: {
          input: todo
        }
      });
    } catch (err) {
      console.log('error creating todo:', err);
    }
  }

  return (
    <div style={styles.container}>
      <Heading level={1}>Hello {user.username}</Heading>
      <Button onClick={signOut} style={styles.button}>
        Sign out
      </Button>
      <h2>Amplify Todos</h2>
      <input
        onChange={(event) => setInput('name', event.target.value)}
        style={styles.input}
        value={formState.name}
        placeholder="Name"
      />
      <input
        onChange={(event) => setInput('description', event.target.value)}
        style={styles.input}
        value={formState.description}
        placeholder="Description"
      />
      <button style={styles.button} onClick={addTodo}>
        Create Todo
      </button>
      {todos.map((todo, index) => (
        <div key={todo.id ? todo.id : index} style={styles.todo}>
          <p style={styles.todoName}>{todo.name}</p>
          <p style={styles.todoDescription}>{todo.description}</p>
        </div>
      ))}
    </div>
  );
};

const styles = {
  container: {
    width: 400,
    margin: '0 auto',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    padding: 20
  },
  todo: { marginBottom: 15 },
  input: {
    border: 'none',
    backgroundColor: '#ddd',
    marginBottom: 10,
    padding: 8,
    fontSize: 18
  },
  todoName: { fontSize: 20, fontWeight: 'bold' },
  todoDescription: { marginBottom: 0 },
  button: {
    backgroundColor: 'black',
    color: 'white',
    outline: 'none',
    fontSize: 18,
    padding: '12px 0px'
  }
};

App.propTypes = {
  signOut: PropTypes.func,
  user: PropTypes.object
}

export default withAuthenticator(App);
```

### Bonus: Use Amplify UI Primitives

You used two Amplify UI components, `Heading` and `Button`. You could also convert the rest of the app to Amplify UI components by replacing the `p` tags with `Text`, the `input`s with `TextField`s and the `div`s with `View`s.

1. Add the `Text`, `TextField`, `View` components to the imported components from Amplify UI:

```javascript
import {
  withAuthenticator,
  Button,
  Heading,
  Text,
  TextField,
  View
} from '@aws-amplify/ui-react';
```

2. Replace the `p` tags with `Text`, the `input`s with `TextField`s and the `div`s with `View`s in your `App` component:

```javascript
<View style={styles.container}>
  <Heading level={1}>Hello {user.username}</Heading>
  <Button style={styles.button} onClick={signOut}>
    Sign out
  </Button>
  <Heading level={2}>Amplify Todos</Heading>
  <TextField
    placeholder="Name"
    onChange={(event) => setInput('name', event.target.value)}
    style={styles.input}
    defaultValue={formState.name}
  />
  <TextField
    placeholder="Description"
    onChange={(event) => setInput('description', event.target.value)}
    style={styles.input}
    defaultValue={formState.description}
  />
  <Button style={styles.button} onClick={addTodo}>
    Create Todo
  </Button>
  {todos.map((todo, index) => (
    <View key={todo.id ? todo.id : index} style={styles.todo}>
      <Text style={styles.todoName}>{todo.name}</Text>
      <Text style={styles.todoDescription}>{todo.description}</Text>
    </View>
  ))}
</View>
```

Using Amplify UI components together makes it easier to manage styling across your entire app.

In this example, you used the Amplify React UI library and the `withAuthenticator` component to quickly get up and running with a real-world authentication flow.

You can also [customize this component](https://ui.docs.amplify.aws/react/components/authenticator/customization) to add or remove fields, update styling, or other configurations. You can even [override function calls](https://ui.docs.amplify.aws/react/components/authenticator/customization#override-function-calls) if needed.

If you would like to build your own authentication flow instead, you can use the [Auth category in Amplify Libraries for JS](/lib/auth/getting-started). Amplify's Auth category class over 20 methods including `signUp`, `signIn`, `forgotPassword`, and `signOut` that allow you full control over all aspects of the user authentication flow. Check out the complete API [here](https://aws-amplify.github.io/amplify-js/api/classes/authclass.html).

In the next section, you'll host your app on the Amplify Console, a hosting service complete with a globally available CDN, automatic deployments, easy custom domains, and CI / CD.
