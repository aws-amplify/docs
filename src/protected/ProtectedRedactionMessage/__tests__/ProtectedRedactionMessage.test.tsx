import * as React from 'react';
import { render } from '@testing-library/react';
import {
  ProtectedRedactionGen1Message,
  ProtectedRedactionGen2Message
} from '../index';
import fs from 'fs';

// REALTIME DATA
const GEN1_V5_REALTIME_DATA_PAGE_PATH =
  'src/pages/gen1/[platform]/prev/build-a-backend/graphqlapi/subscribe-data/index.mdx';

const GEN1_V6_REALTIME_DATA_PAGE_PATH =
  'src/pages/gen1/[platform]/build-a-backend/graphqlapi/subscribe-data/index.mdx';

const GEN2_REALTIME_DATA_PAGE_PATH =
  'src/pages/[platform]/build-a-backend/data/subscribe-data/index.mdx';

// DATA MODELING

const GEN1_V6_DATA_MODELING_PAGE_PATH =
  'src/pages/gen1/[platform]/build-a-backend/graphqlapi/data-modeling/index.mdx';

const GEN2_DATA_MODELING_PAGE_PATH =
  'src/pages/[platform]/build-a-backend/data/data-modeling/relationships/index.mdx';

describe('Protected Redaction Messages', () => {
  /*
        This test is to ensure that the ProtectedRedactionGen1Message component appears on the Gen 1 realtime data pages and cannot be removed or modified without approval.
    */
  it('should render ProtectedRedactionGen1Message component on the Gen 1 V5 realtime data page', async () => {
    const pageData = fs.readFileSync(GEN1_V5_REALTIME_DATA_PAGE_PATH, {
      encoding: 'utf8'
    });
    expect(pageData).toMatch(/<ProtectedRedactionGen1Message \/>/);
  });

  it('should render ProtectedRedactionGen1Message component on the Gen 1 V6 realtime data page', async () => {
    const pageData = fs.readFileSync(GEN1_V6_REALTIME_DATA_PAGE_PATH, {
      encoding: 'utf8'
    });
    expect(pageData).toMatch(/<ProtectedRedactionGen1Message \/>/);
  });

  it('should render ProtectedRedactionGen1Message component on the Gen 2 realtime data page', async () => {
    const pageData = fs.readFileSync(GEN2_REALTIME_DATA_PAGE_PATH, {
      encoding: 'utf8'
    });
    expect(pageData).toMatch(/<ProtectedRedactionGen2Message \/>/);
  });

  it('should render ProtectedRedactionGen1Message component on the Gen 1 V6 data modeling page', async () => {
    const pageData = fs.readFileSync(GEN1_V6_DATA_MODELING_PAGE_PATH, {
      encoding: 'utf8'
    });
    expect(pageData).toMatch(/<ProtectedRedactionGen1Message \/>/);
  });

  it('should render ProtectedRedactionGen1Message component on the Gen 2 data modeling page', async () => {
    const pageData = fs.readFileSync(GEN2_DATA_MODELING_PAGE_PATH, {
      encoding: 'utf8'
    });
    expect(pageData).toMatch(/<ProtectedRedactionGen2Message \/>/);
  });

  /*
      This test is to ensure that the messaging on the ProtectedRedactionGen1Message component does not change
      and cannot be removed or modified without approval.
    */
  it('should render the protected redaction message for Gen 1', async () => {
    const { container } = render(<ProtectedRedactionGen1Message />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the protected redaction message for Gen 2', async () => {
    const { container } = render(<ProtectedRedactionGen2Message />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
