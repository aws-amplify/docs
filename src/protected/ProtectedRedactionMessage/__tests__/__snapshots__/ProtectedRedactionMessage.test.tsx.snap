// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Protected Redaction Messages should render the protected redaction message for Gen 1 1`] = `
<div
  class="amplify-flex amplify-message amplify-message--filled amplify-message--warning"
>
  <div
    class="amplify-flex amplify-message__content"
  >
    <div
      class="amplify-flex"
    >
      <div
        class="amplify-message__icon"
      >
        <svg
          aria-hidden="false"
          aria-label="Warning"
          class="amplify-icon"
          viewBox="0 0 24 24"
        >
          <path
            clip-rule="evenodd"
            d="M1 21H23L12 2L1 21ZM13 18H11V16H13V18ZM13 14H11V10H13V14Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="amplify-message__content"
      >
        <p>
          With versions of Amplify CLI 
          <code>
            @aws-amplify/cli@12.12.2
          </code>
           and API Category
          <code>
            @aws-amplify/amplify-category-api@5.11.5
          </code>
          , an improvement was made to how relational field data is handled in subscriptions when different authorization rules apply to related models in a schema. The improvement redacts the values for the relational fields, displaying them as null or empty, to prevent unauthorized access to relational data. This redaction occurs whenever it cannot be determined that the child model will be protected by the same permissions as the parent model.
        </p>
        <p>
          Because subscriptions are tied to mutations and the selection set provided in the result of a mutation is then passed through to the subscription, relational fields in the result of mutations must be redacted.
        </p>
        <p>
          If an authorized end-user needs access to the redacted relational field they should perform a query to read the relational data.
        </p>
        <p>
          Additionally, subscriptions will inherit related authorization when relational fields are set as required. To better protect relational data, consider modifying the schema to use optional relational fields.
        </p>
        <p>
          Based on the security posture of your application, you can choose to revert to the subscription behavior before this improvement was made.
        </p>
        <p>
          To do so, use the 
          <code>
            subscriptionsInheritPrimaryAuth
          </code>
           feature flag under 
          <code>
            graphqltransformer
          </code>
           in the
           
          <code>
            amplify/backend/cli.json
          </code>
           file.
        </p>
        <ul>
          <li>
            If enabled, subscriptions will inherit the primary model authorization rules for the relational fields.
          </li>
          <li>
            If disabled, relational fields will be redacted in mutation response when there is a difference between auth rules between primary and related models.
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`Protected Redaction Messages should render the protected redaction message for Gen 2 1`] = `
<div
  class="amplify-flex amplify-message amplify-message--filled amplify-message--warning"
>
  <div
    class="amplify-flex amplify-message__content"
  >
    <div
      class="amplify-flex"
    >
      <div
        class="amplify-message__icon"
      >
        <svg
          aria-hidden="false"
          aria-label="Warning"
          class="amplify-icon"
          viewBox="0 0 24 24"
        >
          <path
            clip-rule="evenodd"
            d="M1 21H23L12 2L1 21ZM13 18H11V16H13V18ZM13 14H11V10H13V14Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="amplify-message__content"
      >
        <p>
          With Amplify Data Construct 
          <code>
            @aws-amplify/data-construct@1.8.4
          </code>
          , an improvement was made to how relational field data is handled in subscriptions when different authorization rules apply to related models in a schema. The improvement redacts the values for the relational fields, displaying them as null or empty, to prevent unauthorized access to relational data.
        </p>
        <p>
          This redaction occurs whenever it cannot be determined that the child model will be protected by the same permissions as the parent model.
        </p>
        <p>
          Because subscriptions are tied to mutations and the selection set provided in the result of a mutation is then passed through to the subscription, relational fields in the result of mutations must be redacted.
        </p>
        <p>
          If an authorized end-user needs access to the redacted relational fields, they should perform a query to read the relational data.
        </p>
        <p>
          Additionally, subscriptions will inherit related authorization when relational fields are set as required. To better protect relational data, consider modifying the schema to use optional relational fields.
        </p>
      </div>
    </div>
  </div>
</div>
`;
